firebase-firestore/src/test/java/com/google/firebase/firestore/auth/FirebaseAuthCredentialsProviderTest.java

```diff
@@ -0,0 +1,354 @@
+  }
+
+  @Test
+  public void invalidateTokenShouldCauseGetTokenToForceRefresh() {
+    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
+        new FirebaseAuthCredentialsProvider(new ImmediateDeferred(mockInternalAuthProvider));
+    when(mockGetTokenResult.getToken()).thenReturn("TestToken");
+    when(mockInternalAuthProvider.getAccessToken(anyBoolean()))
+        .thenReturn(Tasks.forResult(mockGetTokenResult));
+
+    firebaseAuthCredentialsProvider.invalidateToken();
+    firebaseAuthCredentialsProvider.getToken();
+
+    verify(mockInternalAuthProvider).getAccessToken(true);
+  }
+
+  @Test
+  public void invalidateTokenShouldOnlyForceRefreshOnTheImmediatelyFollowingGetTokenInvocation() {
+    FirebaseAuthCredentialsProvider firebaseAuthCredentialsProvider =
+        new FirebaseAuthCredentialsProvider(new ImmediateDeferred(mockInternalAuthProvider));
+    when(mockGetTokenResult.getToken()).thenReturn("TestToken");
+    when(mockInternalAuthProvider.getAccessToken(anyBoolean()))
+        .thenReturn(Tasks.forResult(mockGetTokenResult));
+
+    firebaseAuthCredentialsProvider.invalidateToken();
+    firebaseAuthCredentialsProvider.getToken();
+    firebaseAuthCredentialsProvider.getToken();
+
+    verify(mockInternalAuthProvider).getAccessToken(true);
+    verify(mockInternalAuthProvider).getAccessToken(false);
+  }
+
+  /** An implementation of {@link Deferred} whose provider never becomes available. */
+  private static final class UnavailableDeferred implements Deferred<InternalAuthProvider> {
+
+    @Override
+    public void whenAvailable(@NonNull DeferredHandler<InternalAuthProvider> handler) {}
+  }
+
+  /** An implementation of {@link Deferred} whose provider is always available. */
+  private static final class ImmediateDeferred implements Deferred<InternalAuthProvider> {
+
+    private final Provider<InternalAuthProvider> internalAuthProviderProvider;
+
+    ImmediateDeferred(InternalAuthProvider internalAuthProvider) {
+      internalAuthProviderProvider = () -> internalAuthProvider;
+    }
+
+    @Override
+    public void whenAvailable(@NonNull DeferredHandler<InternalAuthProvider> handler) {
+      handler.handle(internalAuthProviderProvider);
+    }
+  }
+
+  /**
+   * An implementation of {@link Deferred} whose provider is initially unavailable, then becomes
+   * available when {@link #setInstance} is invoked.
+   */
+  private static final class DelayedDeferred implements Deferred<InternalAuthProvider> {
+
+    private final Object lock = new Object();
+    private final List<DeferredHandler<InternalAuthProvider>> handlers = new ArrayList<>();
+    private final AtomicReference<Provider<InternalAuthProvider>> internalAuthProviderProviderRef =
+        new AtomicReference<>();
+
+    @Override
+    public void whenAvailable(@NonNull DeferredHandler<InternalAuthProvider> handler) {
+      Provider<InternalAuthProvider> internalAuthProviderProvider;
+      synchronized (lock) {
```

Comment:
Again, I think we should not be mixing `synchronized` and `AtomicReference`. 

Since this is a test class, you could also synchronize on the object itself and use `public synchronized void whenAvailable()` as the method signature.

I also wonder if these methods need to be synchronized at all. Does Mockito invoke the methods from a different thread?